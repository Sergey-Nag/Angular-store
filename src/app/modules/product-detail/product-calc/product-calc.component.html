<div class="card product-calc p-4">
    <div class="row">
        <div class="col">Price, $</div>
        <div class="col-4 text-end">{{price}}</div>
    </div>
    <div class="row align-items-center">
        <div class="col">Count</div>
        <div class="col-4">
            <input
                type="number"
                class="form-control form-control-sm"
                min="0"
                [max]="available"
                [value]="count"
                (change)="handleCountChange($event)">
        </div>
    </div>
    <div class="row">
        <div class="col">
            <b>Total price, $</b>
        </div>
        <div class="col-4 text-end">{{totalPrice}}</div>
    </div>
</div>