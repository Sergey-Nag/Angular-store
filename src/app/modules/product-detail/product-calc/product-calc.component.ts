import { ChangeDetectionStrategy, Component, EventEmitter, Input, OnInit, Output } from '@angular/core';

@Component({
  selector: 'app-product-calc',
  templateUrl: './product-calc.component.html',
  styleUrls: ['./product-calc.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ProductCalcComponent {
  @Input() price: number;
  @Input() available: number;
  @Input() count: number;
  @Output() countChange = new EventEmitter<number>();

  get totalPrice() {
    return (this.price * this.count).toFixed(2);
  }

  onCountChange(e: Event) {
    this.countChange.emit(this.count);
  }
}
