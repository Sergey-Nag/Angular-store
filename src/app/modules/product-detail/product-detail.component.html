<div class="row" *ngIf="book">
    <div class="col-md-8">
        <app-product-info [book]="book"></app-product-info>
    </div>
    <div class="col-md-4 mb-5">
        <app-product-calc
            [price]="book.price"
            [available]="book.count"
            (onCountChange)="handleCountChange($event)">
        </app-product-calc>
        <button
            class="btn btn-primary float-end mt-4"
            [disabled]="count === 0">
            Add to cart
        </button>
    </div>
</div>
<app-spinner *ngIf="isLoading$ | async"></app-spinner>