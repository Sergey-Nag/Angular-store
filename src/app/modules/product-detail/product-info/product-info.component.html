<div class="card p-4">
    <div class="row">
        <div class="col-6">
            <img [src]="book.cover" class="d-block mx-auto" [alt]="book.title">
        </div>
        <div class="col-6">
            <span class="badge bg-info mb-3">{{book.level}}</span>
            <h2>{{book.title}}</h2>
            <span class="text-secondary">{{book.author}}</span>
            <p class="tags mt-3">
                <span 
                    class="badge bg-secondary"
                    *ngFor="let tag of book.tags">
                    {{tag}}
                </span>
            </p>
        </div>
    </div>
    <div class="row mt-3">
        <div class="col-6">
            <p>
                {{book.description}}
            </p>
        </div>
    </div>
</div>