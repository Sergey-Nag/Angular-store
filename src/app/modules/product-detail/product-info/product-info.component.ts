import { ChangeDetectionStrategy, Component, Input, OnInit } from '@angular/core';
import { ParamMap, RouterStateSnapshot } from '@angular/router';
import { Book } from '@shared/models/book.model';

@Component({
  selector: 'app-product-info',
  templateUrl: './product-info.component.html',
  styleUrls: ['./product-info.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ProductInfoComponent implements OnInit {
  @Input() book: Book;

  ngOnInit(): void {
    console.log(this.book);
  }
}
