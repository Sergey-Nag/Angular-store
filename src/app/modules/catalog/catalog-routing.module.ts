import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { CartComponent } from 'src/app/components/cart/cart.component';
import { AuthGuard } from '../auth/guards/auth.guard';
import { CatalogComponent } from './catalog.component';
import { BooksCatalogComponent } from './components/books-catalog/books-catalog.component';
import { ProductDetailComponent } from './components/product-detail/product-detail.component';

const routes: Routes = [
  { path: '', pathMatch: 'full', component: BooksCatalogComponent },
  { path: ':id', component: ProductDetailComponent }
]
@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class CatalogRoutingModule { }