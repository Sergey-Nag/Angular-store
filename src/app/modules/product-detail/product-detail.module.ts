import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { FormsModule } from "@angular/forms";
import { RouterModule, Routes } from "@angular/router";
import { SpinnerComponent } from "@shared/components/spinner/spinner.component";
import { SharedModule } from "@shared/shared.module";
import { ProductCalcComponent } from "./product-calc/product-calc.component";
import { ProductDetailComponent } from "./product-detail.component";
import { ProductInfoComponent } from "./product-info/product-info.component";

const routes: Routes = [
  { path: '', component: ProductDetailComponent}
];

@NgModule({
  declarations: [
    ProductDetailComponent,
    ProductInfoComponent,
    ProductCalcComponent,
  ],
  imports: [
    SharedModule,
    FormsModule,
    RouterModule.forChild(routes)
  ]
})
export class ProductDetailModule {}