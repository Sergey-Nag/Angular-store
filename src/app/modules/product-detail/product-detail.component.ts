import { ChangeDetectionStrategy, ChangeDetectorRef, Component, OnDestroy, OnInit } from '@angular/core';
import { ActivatedRoute, ParamMap } from '@angular/router';
import { HttpLoadingService } from '@core/services/http-loading.service';
import { Book } from '@shared/models/book.model';
import { Observable, Subscription } from 'rxjs';
import { catchError, finalize, switchMap, tap } from 'rxjs/operators';
import { CatalogService } from '../../shared/services/catalog.service';

@Component({
  selector: 'app-product-detail',
  templateUrl: './product-detail.component.html',
  styleUrls: ['./product-detail.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ProductDetailComponent implements OnInit, OnDestroy {
  private paramsSub: Subscription;
  book: Book;
  
  isLoading = this.loading.isLoading;

  constructor(
    private route: ActivatedRoute, 
    private catalog: CatalogService,
    private cdRef: ChangeDetectorRef,
    private loading: HttpLoadingService
  ) {}

  ngOnInit(): void {
    this.paramsSub = this.route.paramMap
      .pipe(
        switchMap((params) => params.get('id')),
        switchMap((id) => this.catalog.getBook(id)),
      )
      .subscribe((book: Book) => {
        this.book = book;

        this.cdRef.detectChanges();
      });
  }

  ngOnDestroy(): void {
    this.paramsSub.unsubscribe();
  }

}
