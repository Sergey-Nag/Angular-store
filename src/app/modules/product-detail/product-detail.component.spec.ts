import { APP_BASE_HREF } from "@angular/common";
import { Book } from "@shared/models/book.model";
import { ProductDetailComponent } from "./product-detail.component";
import { render, RenderResult, screen, waitFor } from '@testing-library/angular';
import { HttpClientTestingModule, HttpTestingController } from "@angular/common/http/testing";
import { AppRoutingModule } from "src/app/app-routing.module";
import { CatalogService } from "@shared/services/catalog.service";
import { of } from "rxjs";
import { ProductInfoComponent } from "./product-info/product-info.component";
import { ProductDetailModule } from "./product-detail.module";
import { ActivatedRoute } from "@angular/router";
import { SharedModule } from "@shared/shared.module";
import { delay } from "rxjs/operators";
import { ProductCalcComponent } from "./product-calc/product-calc.component";
import { RouterTestingModule } from "@angular/router/testing";

describe('Product details', () => {
  let book: Book;
  let rendered: RenderResult<ProductDetailComponent>;

  beforeEach(async () => {
    book = new Book(
      '1',
      6, 
      56.95,
      'Apuntes de Javascript I - Nivel Intermedio',
      'JuanMa Garrido', 
      'Intermediate', 
      'En Castellano) Revision de conceptos (actuales) de javascript desde basicos hasta un nivel intermedio',
      'https://jsbooks.revolunet.com/img/cover-apuntes-javascript-intermedio.png',
      ['core']
    );

    rendered = await render(ProductDetailComponent, {
      imports: [ProductDetailModule, HttpClientTestingModule, AppRoutingModule, SharedModule],
      providers: [
        {
          provide: APP_BASE_HREF,
          useValue: '/'
        },
        {
          provide: ActivatedRoute,
          useValue: {
            snapshot: {
              paramMap: new Map().set('id', book.id)
            }
          }
        },
        {
          provide: CatalogService,
          useValue: {
            getBook: jest.fn().mockReturnValue(of(book).pipe(delay(500)))
          }
        }
      ]
    });

    rendered.navigate(`/catalog/${book.id}`);
  });

  it('Should show the spinner instead content until data is loaded', () => {
    expect(rendered.fixture).toMatchSnapshot('with spinner');
  });

  it('Should show book\'s content instead the spinner after data is loaded', async () => {
    await waitFor(() => screen.getByRole('heading', { name: book.title }), { timeout: 500 });

    expect(rendered.fixture).toMatchSnapshot('with content');
  });
});