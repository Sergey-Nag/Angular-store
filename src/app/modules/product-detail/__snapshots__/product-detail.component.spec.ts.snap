// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product details Should show book's content instead the spinner after data is loaded: with content 1`] = `
<app-product-detail
  __ngContext__={[Function LRootView]}
  book={[Function Book]}
  catalog={[Function Object]}
  cdRef={[Function ViewRef]}
  paramsSub={[Function Subscriber]}
  route={[Function Object]}
>
  <div
    class="row"
  >
    <div
      class="col-md-8"
    >
      <app-product-info>
        <div
          class="card p-4"
        >
          <div
            class="row"
          >
            <div
              class="col-6"
            >
              <img
                alt="Apuntes de Javascript I - Nivel Intermedio"
                class="d-block mx-auto"
                src="https://jsbooks.revolunet.com/img/cover-apuntes-javascript-intermedio.png"
              />
            </div>
            <div
              class="col-6"
            >
              <span
                class="badge bg-info mb-3"
              >
                Intermediate
              </span>
              <h2>
                Apuntes de Javascript I - Nivel Intermedio
              </h2>
              <span
                class="text-secondary"
              >
                JuanMa Garrido
              </span>
              <p
                class="tags mt-3"
              >
                <span
                  class="badge bg-secondary"
                >
                   core 
                </span>
              </p>
            </div>
          </div>
          <div
            class="row mt-3"
          >
            <div
              class="col-6"
            >
              <p>
                 En Castellano) Revision de conceptos (actuales) de javascript desde basicos hasta un nivel intermedio 
              </p>
            </div>
          </div>
        </div>
      </app-product-info>
    </div>
    <div
      class="col-md-4 mb-5"
    >
      <app-product-calc>
        <div
          class="card product-calc p-4"
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              Price, $
            </div>
            <div
              class="col-4 text-end"
            >
              56.95
            </div>
          </div>
          <div
            class="row align-items-center"
          >
            <div
              class="col"
            >
              Count
            </div>
            <div
              class="col-4"
            >
              <input
                class="form-control form-control-sm"
                type="number"
              />
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <b>
                Total price, $
              </b>
            </div>
            <div
              class="col-4 text-end"
            >
              56.95
            </div>
          </div>
        </div>
      </app-product-calc>
      <button
        class="btn btn-primary float-end mt-4"
        disabled=""
      >
        Add to cart
      </button>
    </div>
  </div>
</app-product-detail>
`;

exports[`Product details Should show the spinner instead content until data is loaded: with spinner 1`] = `
<app-product-detail
  __ngContext__={[Function LRootView]}
  book={[Function Book]}
  catalog={[Function Object]}
  cdRef={[Function ViewRef]}
  paramsSub={[Function Subscriber]}
  route={[Function Object]}
>
  <div
    class="row"
  >
    <div
      class="col-md-8"
    >
      <app-product-info>
        <div
          class="card p-4"
        >
          <div
            class="row"
          >
            <div
              class="col-6"
            >
              <img
                alt="Apuntes de Javascript I - Nivel Intermedio"
                class="d-block mx-auto"
                src="https://jsbooks.revolunet.com/img/cover-apuntes-javascript-intermedio.png"
              />
            </div>
            <div
              class="col-6"
            >
              <span
                class="badge bg-info mb-3"
              >
                Intermediate
              </span>
              <h2>
                Apuntes de Javascript I - Nivel Intermedio
              </h2>
              <span
                class="text-secondary"
              >
                JuanMa Garrido
              </span>
              <p
                class="tags mt-3"
              >
                <span
                  class="badge bg-secondary"
                >
                   core 
                </span>
              </p>
            </div>
          </div>
          <div
            class="row mt-3"
          >
            <div
              class="col-6"
            >
              <p>
                 En Castellano) Revision de conceptos (actuales) de javascript desde basicos hasta un nivel intermedio 
              </p>
            </div>
          </div>
        </div>
      </app-product-info>
    </div>
    <div
      class="col-md-4 mb-5"
    >
      <app-product-calc>
        <div
          class="card product-calc p-4"
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              Price, $
            </div>
            <div
              class="col-4 text-end"
            >
              56.95
            </div>
          </div>
          <div
            class="row align-items-center"
          >
            <div
              class="col"
            >
              Count
            </div>
            <div
              class="col-4"
            >
              <input
                class="form-control form-control-sm"
                type="number"
              />
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <b>
                Total price, $
              </b>
            </div>
            <div
              class="col-4 text-end"
            >
              56.95
            </div>
          </div>
        </div>
      </app-product-calc>
      <button
        class="btn btn-primary float-end mt-4"
        disabled=""
      >
        Add to cart
      </button>
    </div>
  </div>
</app-product-detail>
`;
